### PR TITLE
Upgrade to pylint 2.7

### LINKED ISSUE #5435: Upgrade pylint
### What is the expected enhancement?

Upgrade Pylint.

Switching from pylint to flake8 are discussed another issue #1179.

### PR DESCRIPTION
**Pull request by ikkoham**
### Summary

Resolves #5435

### Details and comments

I have a question.

- In `qiskit/transpiler/basepasses.py`, BasePass is hashable but the returning value is None (not integer). Should BasePass be hashable? This is also related to `__eq__`.

### CODE REVIEW

File: qiskit/providers/models/pulsedefaults.py
Comment (ikkoham) [inline qiskit/providers/models/pulsedefaults.py:219]:
```suggestion
            raise AttributeError(f'Attribute {name} is not defined') from ex
```

**Comment by mtreinish** [comment]
@ikkoham thanks for tackling this, I know pylint updates are no fun. But, I'm thinking we might want to put this on hold for now. I was planning to bump our pylint version on the next pylint release to that version. There are issues with all released pylint versions on python 3.9 where it always returns errors. The next pylint release is planned to have support for python 3.9, so I think waiting for that before we merge the updates (just to minimize disruptions). What do you think?

**Comment by ikkoham** [comment]
Thank you for your comment.

> we might want to put this on hold for now.

OK. I understand the situation and your plan. Let's wait until after the next release 2.6.1 to upgrade pylint. However, I think someone should fix the issue about [hashable BasePass](https://github.com/Qiskit/qiskit-terra/pull/5630/files#diff-88bb07327139dc5290c5346139876363194cee51ca0afb7842cfaa9fb0010704L58) and [definition.fset](https://github.com/Qiskit/qiskit-terra/pull/5630/files#diff-f93d8efd2c1c33609fd703738367632df6d709fb84116acc8b511a0d7b94f3cfL128).

**Comment by mtreinish** [comment]
It looks like pylint 2.6.2 still doesn't fix python 3.9 compatibility. Running locally (where python 3.9 is my default version) I still am getting a ton of errors related to type hint changes made in 3.9 that pylint doesn't know how to handle.

**Comment by mtreinish** [comment]
Pylint 2.7.0 was released yesterday and added Python 3.9 support. I've pushed an update to bump this to use 2.7.0 and fixed the errors it found.

**Review by woodsp-ibm — APPROVED** [review]

**Comment by mtreinish** [comment]
> Thank you for your comment.
> 
> > we might want to put this on hold for now.
> 
> OK. I understand the situation and your plan. Let's wait until after the next release 2.6.1 to upgrade pylint. However, I think someone should fix the issue about [hashable BasePass](https://github.com/Qiskit/qiskit-terra/pull/5630/files#diff-88bb07327139dc5290c5346139876363194cee51ca0afb7842cfaa9fb0010704L58) and [definition.fset](https://github.com/Qiskit/qiskit-terra/pull/5630/files#diff-f93d8efd2c1c33609fd703738367632df6d709fb84116acc8b511a0d7b94f3cfL128).

I think this is ready to go now that 2.7.0 has been released and adds python 3.9 support.

As for the `__hash__` function it's a weird edge case. I believe the `BasePass` class is supposed to be an abstract class that isn't instantiated on it's own. That being said looking through the code nothing is overloading the `__hash__` method. This does look like a bug since we shouldn't be returning `None` from `__hash__` as it's not a valid hash. Looking through the git history it was added in: https://github.com/Qiskit/qiskit-terra/pull/1968 which doesn't really provide sufficient context for me to figure out what it was trying to do. My feeling is we should just remove the `__hash__` method there, but maybe we should do that in a separate PR so we can discuss why it was added.

**Review by manoelmarques — DISMISSED** [review]

**Comment by levbishop** [comment]
I think the __hash__ stuff might be  working in practice, because the metaclass overrides the _hash value, but its worth taking a close look rather than just adding the pylint suppression.

File: requirements-dev.txt
Comment (ikkoham) [inline requirements-dev.txt:10]:
```suggestion
astroid==2.5
pylint==2.7.1
```

**Comment by ikkoham** [comment]
Thanks. I have updated pylint from `2.7.0` to `2.7.1`.

`__hash__ = None` may be correct, but it should be done in follow up issue.

**Review by levbishop — APPROVED** [review]
Thanks for doing this. Can you open another issue to keep track of the hashing situation?

**Review by manoelmarques — APPROVED** [review]

**Review by manoelmarques — APPROVED** [review]

**Review by levbishop — APPROVED** [review]

**Review by mtreinish — APPROVED** [review]

**Pull request merged**
